src/ximea_ros_cam.cpp

```diff
@@ -869,9 +767,303 @@ void XimeaROSCam::triggerCb(const std_msgs::Empty::ConstPtr& msg) {
+
+    // ROS_INFO_STREAM("Timestamps:" << // TODO delete or demote to DEGUG
+    //         " System: " << timestamp_queue_.back().first << 
+    //         " Camera: " << timestamp_queue_.back().second );
+
+    // monitor round trip speed
+    ros::Duration round_trip_duration(post_stamp - pre_stamp);
+    if(round_trip_duration > ros::Duration(0.0025)) // arbitary 2.5 ms (I see a 2 ms average with an xiC on my system)
+        ROS_WARN_STREAM_THROTTLE(3, "Camera timestamp query took " << round_trip_duration << " seconds." );    
+}
+
+ros::Time XimeaROSCam::iterpolateTimestamp(const XI_IMG& frame){
+    if(!camera_timestamp_supported_)  
+        return ros::Time::now();
+
+    sampleCameraTimestamp();
+
+    std::deque< std::pair<ros::Time,ros::Duration> >& queue = timestamp_queue_; // alias 
+    if(queue.size() < 2)
+    {  
+        ROS_ERROR("Not enough timestamp samples to interpolate.");
+        return ros::Time::now();
+    }
+
+    ros::Duration camera_stamp(frame.tsSec, frame.tsUSec * 1000);
+    // TODO adjust camera stamp to be the start of exposure (model dependant)
+    // consider using a precalculated bool to determin if exposure time should
+    // be deducted  
+    
+    bool clock_wrap = queue.front().second > queue.back().second; 
+    if(clock_wrap)
+    {
+        ROS_WARN("Clock wrap detected. Timestamp handling experimetal!");
+        // xiQ, xiD 
+        // 40 bit microsecond number - (overlaps after 305 hours)
+        // TimeStamp is recorded at the start of Data Readout 
+
+        // xiC, xiB, xiX, xiT 
+        // 64 bit 4-nanoseconds number (overlaps after 2339 years) (sic)
+        // TimeStamp is recorded at the start of Exposure.
+        // It is assumed that these cameras use the full tsSec value 
+
+        // untested experimental handling!
+        ros::Duration wrap_period;
+        ros::Duration rough_wrap_period = 
+                (queue.front().second - queue.back().second) + 
+                (queue.back().first - queue.front().first);
+
+        const uint64_t million = 1000000;
+        const uint64_t forty_bits = uint64_t(1) << 40;
+        const ros::Duration forty_bit_period(forty_bits / million, forty_bits % million);
+        const ros::Duration ros_duration_period(1<<31,0); // ros duration overflow
+
+        if((rough_wrap_period - forty_bit_period).sec == 0) 
+            wrap_period = forty_bit_period; // ~305 hours
+        else if((rough_wrap_period - ros_duration_period).sec == 0)  
+            wrap_period = ros::Duration(1<<31,0); // ~68 years (worlds longest running camera)
+        else
+            wrap_period = rough_wrap_period;
+         
+        ROS_WARN_STREAM("Adjusting timestamps using a period of " << wrap_period);
+        using iterator = std::deque< std::pair<ros::Time,ros::Duration> >::iterator;
+        for(iterator it = queue.begin(); it != queue.end(); ++it)
+            if(it->second > queue.back().second)
```

Comment:
It is expected behaviour here, once `queue.back().second` reaches the wraparound (which means that `queue.back().second` reaches `~0`), then we will subtract *almost all* of the previous values by the wrap around?

For example:
If we have a camera timer which wraps at `1.0`, and we obtain timestamps every `0.2`, with the first timestamp obtained at `0.1`, then our first 6 timestamps will be:

`0.1, 0.3, 0.5, 0.7, 0.9, 0.1`

With this check, it means our result should be (with the `it->second -= wrap_period` statement):

`0.1, -0.7, -0.5, -0.3, 0.1`

which puts most of the timestamps in the negatives, as well as an edge case where the first (front) timestamp being still in the positive.

Reply:
It is expected the the camera clock will not overflow twice during the period covered by the deque.
It is expected that when the wrap gets detected only the last pair in the deque will contain a wrapped stamp, this is because other than initialising we only sample the camera clock in the interpolation function where we correct previous stamps to be in the new time frame. The stamp from XI_IMAGE may or may not be wrapped and I realise there is a bug here where we correct the stamps in one frame but might still get images in with non wrapped stamps in subsequent frames. I guess another check could catch this.

Reply:
I see what you mean here, I think I was reading through this in sequence and did not take into account of

`while(queue.size() > 2 && queue[1].second < camera_stamp) { queue.pop_front(); }`

The only scenario where I feel we will get a growing `queue` size is when the camera capture continuously times out and somehow the `camera_stamp` does not get updated.

Assuming that this is not the case and we will know when the camera is not behaving properly if there are timeouts, then you are right that the camera clock will not overflow twice.

I think just by the fact that the timestamps will never surpass the max, we will be working with negative timestamps in `queue[0].second` and `camera_stamp` for our interpolation since we are subtracting these with `wrap_period`.

This is fine though since the interpolation scalar computation is a proportional computation - so we should be good in the expected case.

As for the edge case where the image timestamp is not corrected (maybe this will only happen at a high frequency asynchronous scenario), good catch!